src/github.com/matrix-org/go-neb/services/circleci/circleci.go

```diff
@@ -0,0 +1,270 @@
+func notifToTemplate(n WebhookNotification) map[string]string {
+	t := make(map[string]string)
+	//Get Payload to variable
+	p := n.Payload
+	t["repository_slug"] = p.Username + "/" + p.Reponame
+	t["repository"] = t["repository_slug"] // Deprecated form but still used everywhere in people's templates
+	t["repository_name"] = p.Reponame
+	t["build_number"] = strconv.Itoa(p.BuildNum)
+	t["build_id"] = t["build_number"] // CircleCI doesn't have a difference between number and ID but to be consistent with TravisCI
+	shaLength := len(p.VcsRevision)
+	if shaLength > 10 {
+		shaLength = 10
+	}
+	t["commit"] = p.VcsRevision[:shaLength] // shortened commit SHA
+	t["author"] = p.CommitterName      // author: commit author name
+	// commit_message: commit message of build
+	// commit_subject: first line of the commit message
+	t["commit_message"] = p.Body
+	subjAndMsg := strings.SplitN(p.Body, "\n", 2)
+	t["commit_subject"] = subjAndMsg[0]
+	if p.Status != "" {
+		t["result"] = p.Status
+	}
+	t["message"] = p.Outcome // message: CircleCI message to the build
+
+	if !p.StartTime.IsZero()  && !p.StopTime.IsZero() {
+			t["duration"] = p.StopTime.Sub(p.StartTime).String()
+			t["elapsed_time"] = t["duration"]
+	}
+
+	t["build_url"] = p.BuildURL
+
+	// Make the full struct data into the Map without manual mapping
+	rawT := structs.Map(p)
+	for key, value := range rawT {
+		switch value := value.(type) {
+		case string:
+			t[CamelCaseToUnderscore(key)] = value
+		case int:
+			t[CamelCaseToUnderscore(key)] = strconv.Itoa(value)
+		case time.Time:
+			t[CamelCaseToUnderscore(key)] = value.String()
+		}
+	}
+	return t
+}
+
+func outputForTemplate(circleciTmpl string, tmpl map[string]string) (out string) {
+	if circleciTmpl == "" {
+		circleciTmpl = DefaultTemplate
+	}
+	out = circleciTmpl
+	for tmplVar, tmplValue := range tmpl {
+		out = strings.Replace(out, "%{"+tmplVar+"}", tmplValue, -1)
+	}
+	return out
+}
+
+// OnReceiveWebhook receives requests from CircleCI and possibly sends requests to Matrix as a result.
+//
+// If the repository matches a known Git* repository, a notification will be formed from the
+// template for that repository and a notice will be sent to Matrix.
+//
+// Go-NEB cannot register with CircleCI for webhooks automatically. The user must manually add the
+// webhook endpoint URL to their .circleci/config.yml file:
+//    notify:
+//		webhooks:
+//			- url: https://example.com/services/hooks/circle
+//
+// See https://circleci.com/docs/1.0/configuration/#notify for more information.
+func (s *Service) OnReceiveWebhook(w http.ResponseWriter, req *http.Request, cli *gomatrix.Client) {
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"Body":  req.Body,
+			log.ErrorKey: err,
+		}).Warn("Failed to Read Body")
+		w.WriteHeader(403)
+		return
+	}
+	//fmt.Printf("%s\n", body)
+
+	var notif WebhookNotification
+	if err := json.Unmarshal(body, &notif); err != nil {
```

Comment:
You can replace this and the `ioutil` call with a `err = json.NewDecoder(req.Body).Decode(&notify)` as per https://stackoverflow.com/a/21198571/3532235